### PR TITLE
Fix packages.config to project mapping in NuGet.exe

### LINKED ISSUE #13456: ##[error]The nuget command failed with exit code(1) and error(An item with the same key has already been added.
### NuGet Product Used

NuGet.exe

### Product Version

6.10

### Worked before?

6.9

### Impact

I'm unable to use this version

### Repro Steps & Context

In our yaml file we restore NuGet packages from our own feed and from NuGet.org. 
We are specifying Version 6.9.1 as a workaround.

image

steps:
- task: NuGetToolInstaller@1

- task: NuGetCommand@2
  displayName: "Restore NuGet package dependencies from custom name feed."
  inputs:
    command: 'restore'
    restoreSolution: '$(solution)'
    feedsToUse: 'select'
    vstsFeed: 'our feed key'

- task: NuGetCommand@2
  displayName: "Restore NuGet package dependencies from NuGet.org feed."
  inputs:
    command: 'restore'
    restoreSolution: '**/*.sln'
    feedsToUse: 'select'

### Verbose Logs

```shell
2024-05-21T19:19:32.0057280Z ##[section]Starting: Restore NuGet package dependencies from feed.
2024-05-21T19:19:32.0067250Z ==============================================================================
2024-05-21T19:19:32.0067384Z Task         : NuGet
2024-05-21T19:19:32.0067449Z Description  : Restore, pack, or push NuGet packages, or run a NuGet command. Supports NuGet.org and authenticated feeds like Azure Artifacts and MyGet. Uses NuGet.exe and works with .NET Framework apps. For .NET Core and .NET Standard apps, use the .NET Core task.
2024-05-21T19:19:32.0067720Z Version      : 2.238.1
2024-05-21T19:19:32.0067779Z Author       : Microsoft Corporation
2024-05-21T19:19:32.0067850Z Help         : https://docs.microsoft.com/azure/devops/pipelines/tasks/package/nuget
2024-05-21T19:19:32.0067952Z ==============================================================================
2024-05-21T19:19:33.0007158Z [command]C:\Windows\system32\chcp.com 65001
2024-05-21T19:19:33.0964601Z Active code page: 65001
2024-05-21T19:19:33.4640790Z Detected NuGet version 6.10.0.107 / 6.10.0+6970ffe10efff51d43862ea8f8a2ef0170250262.6970ffe10efff51d43862ea8f8a2ef0170250262
2024-05-21T19:19:33.5217240Z [command]C:\hostedtoolcache\windows\NuGet\6.10.0\x64\nuget.exe sources Add -NonInteractive -Name eb5226ba-ad7e-4568-ab21-ae78862f0a66 -Source https://pkgs.dev.azure.com/repo-name/feed-key-part-1/_packaging/feed-key-part-2/nuget/v3/index.json -ConfigFile D:\a\1\Nuget\tempNuGet_1007.config
2024-05-21T19:19:35.7583675Z Package source with Name: eb5226ba-ad7e-4568-ab21-ae78862f0a66 added successfully.
2024-05-21T19:19:35.7685511Z [command]C:\hostedtoolcache\windows\NuGet\6.10.0\x64\nuget.exe sources Add -NonInteractive -Name NuGetOrg -Source https://api.nuget.org/v3/index.json -ConfigFile D:\a\1\Nuget\tempNuGet_1007.config
2024-05-21T19:19:36.6715155Z Package source with Name: NuGetOrg added successfully.
2024-05-21T19:19:36.6814512Z [command]C:\hostedtoolcache\windows\NuGet\6.10.0\x64\nuget.exe restore D:\a\1\s\Portal.sln -Verbosity Detailed -NonInteractive -ConfigFile D:\a\1\Nuget\tempNuGet_1007.config
2024-05-21T19:19:54.1074961Z An item with the same key has already been added.
2024-05-21T19:19:54.1075287Z System.ArgumentException: An item with the same key has already been added.
2024-05-21T19:19:54.1075522Z    at System.ThrowHelper.ThrowArgumentException(ExceptionResource resource)
2024-05-21T19:19:54.1075787Z    at System.Collections.Generic.Dictionary`2.Insert(TKey key, TValue value, Boolean add)
2024-05-21T19:19:54.1076057Z    at NuGet.CommandLine.RestoreCommand.GetPackagesConfigToProjectPath(PackageRestoreInputs packageRestoreInputs)
2024-05-21T19:19:54.1076340Z    at NuGet.CommandLine.RestoreCommand.<PerformNuGetV2RestoreAsync>d__59.MoveNext()
2024-05-21T19:19:54.1076573Z --- End of stack trace from previous location where exception was thrown ---
2024-05-21T19:19:54.1076811Z    at System.Runtime.ExceptionServices.ExceptionDispatchInfo.Throw()
2024-05-21T19:19:54.1077080Z    at System.Runtime.CompilerServices.TaskAwaiter.HandleNonSuccessAndDebuggerNotification(Task task)
2024-05-21T19:19:54.1077340Z    at NuGet.CommandLine.RestoreCommand.<ExecuteCommandAsync>d__52.MoveNext()
2024-05-21T19:19:54.1077587Z --- End of stack trace from previous location where exception was thrown ---
2024-05-21T19:19:54.1077808Z    at System.Runtime.ExceptionServices.ExceptionDispatchInfo.Throw()
2024-05-21T19:19:54.1078068Z    at System.Runtime.CompilerServices.TaskAwaiter.HandleNonSuccessAndDebuggerNotification(Task task)
2024-05-21T19:19:54.1078287Z    at NuGet.CommandLine.Command.Execute()
2024-05-21T19:19:54.1078509Z    at NuGet.CommandLine.Program.MainCore(String workingDirectory, String[] args)
2024-05-21T19:19:54.1078701Z NuGet Version: 6.10.0.107
2024-05-21T19:19:54.1079022Z MSBuild auto-detection: using msbuild version '16.11.2.50704' from 'C:\Program Files (x86)\Microsoft Visual Studio\2019\Enterprise\MSBuild\Current\bin'. Use option -MSBuildVersion to force nuget to use a specific version of MSBuild.
2024-05-21T19:19:54.1081182Z WARNING: Project file D:\a\1\s\Portal\ cannot be found.
2024-05-21T19:19:54.1081359Z MSBuild P2P timeout [ms]: 120000
2024-05-21T19:19:54.1082148Z C:\Program Files (x86)\Microsoft Visual Studio\2019\Enterprise\MSBuild\Current\bin\msbuild.exe "C:\Users\VssAdministrator\AppData\Local\Temp\NuGetScratch\z4wkm3zx.x40.nugetinputs.targets" /t:GenerateRestoreGraphFile /nologo /nr:false /v:q /p:NuGetPropsFile="C:\Users\VssAdministrator\AppData\Local\Temp\NuGetScratch\qmszf1y4.ohy.nugetrestore.props" /p:NuGetRestoreTargets="C:\Users\VssAdministrator\AppData\Local\Temp\NuGetScratch\10k25o4w.k4a.nugetrestore.targets" /p:RestoreUseCustomAfterTargets="True" /p:DisableCheckingDuplicateNuGetItems="True" /p:RestoreTaskAssemblyFile="C:\hostedtoolcache\windows\NuGet\6.10.0\x64\nuget.exe" /p:RestoreSolutionDirectory="D:\a\1\s\\" /p:RestoreConfigFile="D:\a\1\Nuget\tempNuGet_1007.config" /p:SolutionDir="D:\a\1\s\\" /p:SolutionName="Portal"
2024-05-21T19:19:54.1082780Z 
2024-05-21T19:19:54.1847615Z ##[error]The nuget command failed with exit code(1) and error(An item with the same key has already been added.
System.ArgumentException: An item with the same key has already been added.
   at System.ThrowHelper.ThrowArgumentException(ExceptionResource resource)
   at System.Collections.Generic.Dictionary`2.Insert(TKey key, TValue value, Boolean add)
   at NuGet.CommandLine.RestoreCommand.GetPackagesConfigToProjectPath(PackageRestoreInputs packageRestoreInputs)
   at NuGet.CommandLine.RestoreCommand.<PerformNuGetV2RestoreAsync>d__59.MoveNext()
--- End of stack trace from previous location where exception was thrown ---
   at System.Runtime.ExceptionServices.ExceptionDispatchInfo.Throw()
   at System.Runtime.CompilerServices.TaskAwaiter.HandleNonSuccessAndDebuggerNotification(Task task)
   at NuGet.CommandLine.RestoreCommand.<ExecuteCommandAsync>d__52.MoveNext()
--- End of stack trace from previous location where exception was thrown ---
   at System.Runtime.ExceptionServices.ExceptionDispatchInfo.Throw()
   at System.Runtime.CompilerServices.TaskAwaiter.HandleNonSuccessAndDebuggerNotification(Task task)
   at NuGet.CommandLine.Command.Execute()
   at NuGet.CommandLine.Program.MainCore(String workingDirectory, String[] args))
2024-05-21T19:19:54.1858650Z ##[error]Packages failed to restore
2024-05-21T19:19:54.1868463Z ##[section]Finishing: Restore NuGet package dependencies from  feed.
```

### PR DESCRIPTION
**Pull request by nkolev92**
## Bug

Fixes: https://github.com/NuGet/Home/issues/13456

Regression? Last working version:

## Description

The implementation for PC vulnerability checking did not account for the fact that multiple projects might be in the same folder and lead to duplicate packages.config to project mapping. 

This fixes that and adds test cases in both nuget.exe restore and msbuild -t:restore. 

I've add some in  line comments to aid with reviewing.

Please note that this will be patched in 6.10, so take that into consideration when reviewing (ie. focus on correctness, extra clean-up can be done as follow-up and review the tests for correctness).

## PR Checklist

- [x] PR has a meaningful title
- [x] PR has a linked issue.
- [x] Described changes

- **Tests**
  - [x] Automated tests added
  - **OR**
  
  - [ ] Test exception
  - **OR**
  - [ ] N/A 

- **Documentation**
  
  - [ ] Documentation PR or issue filled
  - **OR**
  - [x] N/A

### CODE REVIEW

File: src/NuGet.Clients/NuGet.CommandLine/Commands/RestoreCommand.cs
Comment (nkolev92) [inline src/NuGet.Clients/NuGet.CommandLine/Commands/RestoreCommand.cs:282]: Ensure we don't process the same config twice. 

There's more performant ways of doing but:

1. It's NuGet.exe, so this cost is not everywhere and very minor.
2. More code = more risk, potentially more challenges getting approval.

File: src/NuGet.Clients/NuGet.CommandLine/Commands/RestoreCommand.cs
Comment (nkolev92) [inline src/NuGet.Clients/NuGet.CommandLine/Commands/RestoreCommand.cs:279]: While multiple projects to 1 packages.config is not recommended, it is obviously something that does happen, and I don't want to rock the boat in a patch release, so the fix is to just handle that scenario.

https://github.com/NuGet/Client.Engineering/issues/2851 covers more.

**Review by martinrrm — DISMISSED** [review]

File: test/NuGet.Clients.Tests/NuGet.CommandLine.Test/NuGetRestoreCommandTest.cs
Comment (zivkan) [inline test/NuGet.Clients.Tests/NuGet.CommandLine.Test/NuGetRestoreCommandTest.cs:3629]: the stop should be in a finally block, so that if `CommandRunner.Run` throws, the mock server is still stopped.

File: test/NuGet.Core.FuncTests/Msbuild.Integration.Test/MsbuildRestoreTaskTests.cs
Comment (zivkan) [inline test/NuGet.Core.FuncTests/Msbuild.Integration.Test/MsbuildRestoreTaskTests.cs:1608]: Same as other comment, the mock server needs to be stopped, even when an exception is thrown.

**Review by zivkan — DISMISSED** [review]

File: test/NuGet.Clients.Tests/NuGet.CommandLine.Test/NuGetRestoreCommandTest.cs
Comment (nkolev92) [reply test/NuGet.Clients.Tests/NuGet.CommandLine.Test/NuGetRestoreCommandTest.cs:3629]: The server is disposable, and a using statement it used. 

dispose calls stop.

File: test/NuGet.Core.FuncTests/Msbuild.Integration.Test/MsbuildRestoreTaskTests.cs
Comment (nkolev92) [reply test/NuGet.Core.FuncTests/Msbuild.Integration.Test/MsbuildRestoreTaskTests.cs:1608]: Same as the previous. Dispose calls Stop as well.

**Comment by nkolev92** [comment]
Blocked by some restore issues.
Fix for that in https://github.com/NuGet/NuGet.Client/pull/5827.

**Review by zivkan — APPROVED** [review]

**Pull request merged**
